pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +449,705 @@ pub mod pallet {
 				false
+			};
+			self.reset_bottom_data::<T>(&bottom_delegations);
+			<BottomDelegations<T>>::insert(candidate, bottom_delegations);
+			Ok(in_top_after)
+		}
+		/// Decrease delegation
+		pub fn decrease_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+			bond: Balance,
+			less: BalanceOf<T>,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let lowest_top_eq_highest_bottom =
+				self.lowest_top_delegation_amount == self.highest_bottom_delegation_amount;
+			let bond_geq_lowest_top = bond >= self.lowest_top_delegation_amount;
+			let delegation_dne_err: DispatchError = Error::<T>::DelegationDNE.into();
+			if bond_geq_lowest_top && !lowest_top_eq_highest_bottom {
+				// definitely in top
+				self.decrease_top_delegation::<T>(candidate, delegator.clone(), bond, less)
+			} else if bond_geq_lowest_top && lowest_top_eq_highest_bottom {
+				// update top but if error then update bottom (because could be in bottom because
+				// lowest_top_eq_highest_bottom)
+				let result =
+					self.decrease_top_delegation::<T>(candidate, delegator.clone(), bond, less);
+				if result == Err(delegation_dne_err) {
+					self.decrease_bottom_delegation::<T>(candidate, delegator, less)
+				} else {
+					result
+				}
+			} else {
+				self.decrease_bottom_delegation::<T>(candidate, delegator, less)
 			}
 		}
-		/// Return true if in_top after call
-		pub fn decrease_delegation(&mut self, delegator: A, less: B) -> bool {
-			let mut in_top = false;
-			let mut new_lowest_top: Option<Bond<A, B>> = None;
-			for x in &mut self.top_delegations {
-				if x.owner == delegator {
-					x.amount -= less;
-					// if there is at least 1 delegator in bottom delegators, compare it to check
-					// if it should be swapped with lowest top delegation and put in top
-					// >pop requires push later on to reset in case it isn't used
-					if let Some(highest_bottom) = self.bottom_delegations.pop() {
-						if highest_bottom.amount > x.amount {
-							new_lowest_top = Some(highest_bottom);
-						} else {
-							// >required push to reset self.bottom_delegations
-							self.bottom_delegations.push(highest_bottom);
+		/// Decrease top delegation
+		pub fn decrease_top_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+			bond: Balance,
+			less: BalanceOf<T>,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let bond_after_less_than_highest_bottom =
+				bond - less.into() < self.highest_bottom_delegation_amount;
+			let full_top_and_nonempty_bottom = matches!(self.top_capacity, CapacityStatus::Full)
+				&& !matches!(self.bottom_capacity, CapacityStatus::Empty);
+			let mut top_delegations =
+				<TopDelegations<T>>::get(candidate).ok_or(Error::<T>::CandidateDNE)?;
+			let in_top_after =
+				if bond_after_less_than_highest_bottom && full_top_and_nonempty_bottom {
```

Comment:
These are both a mouthful of a variable name, some comments would help a bit (esp. the first)

Reply:
I'll add a comment. The first one means that the delegation after the `decrease-delegation` will be strictly less than the highest bottom delegation. The second one means the top delegations is full and the bottom delegations has at least one delegation.